cmd/ocsp-updater/main.go

```diff
@@ -48,26 +113,97 @@ func setupClients(c cmd.Config, stats statsd.Statter) (rpc.CertificateAuthorityC
 	cmd.FailOnError(err, "Unable to create CA client")
 
-	return cac, closeChan
+	pubRPC, err := rpc.NewAmqpRPCClient("OCSP->Publisher", c.AMQP.Publisher.Server, ch, stats)
+	cmd.FailOnError(err, "Unable to create RPC client")
+
+	pubc, err := rpc.NewPublisherClient(pubRPC)
+	cmd.FailOnError(err, "Failed to create Publisher client")
+
+	return cac, pubc, closeChan
 }
 
-func (updater *OCSPUpdater) processResponse(tx *gorp.Transaction, serial string) error {
-	certObj, err := tx.Get(core.Certificate{}, serial)
-	if err != nil {
-		return err
+var errNoStaleResponses = errors.New("No stale responses to update")
+
+func (updater *OCSPUpdater) findStaleOCSPResponses(oldestLastUpdatedTime time.Time) ([]core.Certificate, error) {
+	var certs []core.Certificate
+	_, err := updater.dbMap.Select(
+		&certs,
+		`SELECT cert.*
+		 FROM certificateStatus AS cs
+		 JOIN certificates AS cert
+		 ON cs.serial = cert.serial
+		 WHERE cs.ocspLastUpdated < :lastUpdate
+		 AND cert.expires > now()
+		 ORDER BY cs.ocspLastUpdated ASC
+		 LIMIT :limit`,
+		map[string]interface{}{
+			"lastUpdate": oldestLastUpdatedTime,
+			"limit":      updater.oldOCSPBatch,
+		},
+	)
+	if err == sql.ErrNoRows || len(certs) == 0 {
+		return certs, errNoStaleResponses
 	}
-	statusObj, err := tx.Get(core.CertificateStatus{}, serial)
-	if err != nil {
-		return err
+	return certs, err
+}
+
+func (updater *OCSPUpdater) getNumberOfReceipts(serial string) (int, error) {
+	var count int
+	err := updater.dbMap.SelectOne(
+		&count,
+		"SELECT COUNT(*) FROM sctReceipts WHERE certificateSerial = :serial",
+		map[string]interface{}{"serial": serial},
+	)
+	return count, err
+}
+
+var errNoNewCertificates = errors.New("No certificates issued since last check")
+
+func (updater *OCSPUpdater) getCertificatesIssuedSince(since time.Time, batchSize int) ([]core.Certificate, error) {
+	var certs []core.Certificate
+	_, err := updater.dbMap.Select(
+		&certs,
+		`SELECT * FROM certificates
+		 WHERE issued > :since
+		 LIMIT :limit`,
+		map[string]interface{}{
+			"since": since,
+			"limit": batchSize,
+		},
+	)
+	if err == sql.ErrNoRows || len(certs) == 0 {
+		return certs, errNoNewCertificates
 	}
+	return certs, err
+}
 
-	cert, ok := certObj.(*core.Certificate)
-	if !ok {
-		return fmt.Errorf("Cast failure")
+func (updater *OCSPUpdater) getSerialsIssuedSince(since time.Time, batchSize int) ([]string, error) {
+	var serials []string
+	_, err := updater.dbMap.Select(
+		&serials,
+		`SELECT serial FROM certificates
+		 WHERE issued > :since
+		 LIMIT :limit`,
+		map[string]interface{}{
+			"since": since,
+			"limit": batchSize,
+		},
+	)
+	if err == sql.ErrNoRows || len(serials) == 0 {
+		return serials, errNoNewCertificates
 	}
-	status, ok := statusObj.(*core.CertificateStatus)
-	if !ok {
-		return fmt.Errorf("Cast failure")
+	return serials, err
+}
+
+func (updater *OCSPUpdater) generateResponse(tx *gorp.Transaction, cert core.Certificate) error {
```

Comment:
Calling an RPC while holding a transaction implies we need consistent results which we can't guarantee (this is the same as the CA serial id problem we had). Also, while the default transaction timeout is currently very high, we will be working to reduce that and the query timeout, as well, so the transaction is being held for `N*<mean of the RPC time>` which can be long.

Instead, can we request the signatures from the CA in one loop and then loop through them to write the new OCSPResponses?